ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Protocol.hs

```diff
@@ -195,45 +196,47 @@ forgeTPraosFields updateNodeState TPraosProof{..} kesPeriod mkToSign = do
 evolveKESKeyIfNecessary
   :: forall m c. (MonadRandom m, TPraosCrypto c)
   => NodeState.Update m (TPraosNodeState c)
-  -> SL.KESPeriod
+  -> SL.KESPeriod -- ^ Relative KES period (to the start period of the OCert)
   -> m (SignKeyKES (KES c))
 evolveKESKeyIfNecessary updateNodeState (SL.KESPeriod kesPeriod) = do
     getOudatedKeyOrCurrentKey >>= \case
       Right currentKey -> return currentKey
       Left outdatedKey -> do
-        newKey <- evolveKey outdatedKey
+        let newKey@(HotKey _ key) = evolveKey outdatedKey
         saveNewKey newKey
-        return newKey
+        return key
   where
-    -- | Return either (@Left@) an outdated key (setting the node state to
-    -- 'TPraosKeyEvolving') or (@Right@) a key that's up-to-date w.r.t. the
-    -- current KES period (leaving the node state to 'TPraosKeyAvailable').
+    -- | Return either (@Left@) an outdated key with its current period (setting
+    -- the node state to 'TPraosKeyEvolving') or (@Right@) a key that's
+    -- up-to-date w.r.t. the current KES period (leaving the node state to
+    -- 'TPraosKeyAvailable').
     getOudatedKeyOrCurrentKey
-      :: m (Either (SignKeyKES (KES c)) (SignKeyKES (KES c)))
+      :: m (Either (HotKey c) (SignKeyKES (KES c)))
     getOudatedKeyOrCurrentKey = NodeState.runUpdate updateNodeState $ \case
       TPraosKeyEvolving ->
         -- Another thread is currently evolving the key; wait
         Nothing
-      TPraosKeyAvailable key
-        | let kesPeriodOfKey = KES.currentPeriodKES () key
-        , kesPeriodOfKey < kesPeriod
+      TPraosKeyAvailable hk@(HotKey kesPeriodOfKey key)
+        | kesPeriodOfKey < kesPeriod
           -- Must evolve key
-        -> return (TPraosKeyEvolving, Left key)
+        -> return (TPraosKeyEvolving, Left hk)
         | otherwise
-        -> return (TPraosKeyAvailable key, Right key)
+        -> return (TPraosKeyAvailable hk, Right key)
       TPraosNoKey ->
         error "no KES key available"
 
     -- | Evolve the given key so that its KES period matches @kesPeriod@.
-    evolveKey :: SignKeyKES (KES c) -> m (SignKeyKES (KES c))
-    evolveKey outdatedKey =
-      case KES.updateKES () outdatedKey kesPeriod of
-        -- TODO
-        Nothing         -> error "Could not update KES key"
-        Just evolvedKey -> return evolvedKey
+    evolveKey :: HotKey c -> HotKey c
+    evolveKey (HotKey oldPeriod outdatedKey) = go outdatedKey oldPeriod kesPeriod
+      where
+        go !_ c t | t < c = error "Asked to evolve KES key to old period"
+        go !sk c t | c == t = HotKey kesPeriod sk
+        go !sk c t = case KES.updateKES () sk c of
+          Nothing  -> error "Could not update KES key"
+          Just sk' -> go sk' (c + 1) t
```

Comment:
I find the following a bit easier to read:
```suggestion
        go !sk c t
          | t < c
          = error "Asked to evolve KES key to old period"
          | c == t
          = HotKey kesPeriod sk
          | otherwise
          = case KES.updateKES () sk c of
              Nothing  -> error "Could not update KES key"
              Just sk' -> go sk' (c + 1) t
```